### PR TITLE
Async deleteFiles and cleanup

### PR DESCRIPTION
**Pull request by jrouault**
deleteFiles is done in the background, deleteFile is not.

poke @ronkorving @micky2be @wizcorp-olivier for potential reviews.

### CODE REVIEW

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java [hunk 1]
```diff
@@ -96,18 +98,18 @@ public JSONObject getAllAssets() {
...
+
         // Open the .db file from assets directory
         try {
             InputStream is = that.getAssets().open(DATABASE_INTERNAL_FILE_PATH + DATABASE_NAME);
 
             // Copy the database into the destination
             OutputStream os = new FileOutputStream(DATABASE_EXTERNAL_FILE_PATH + File.separator + DATABASE_NAME);
-        
+
             byte[] buffer = new byte[1024];
```
Comment (ronkorving) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: Unrelated to this PR, but just noticed... Isn't this a really inefficiently small buffer size for doing disk I/O?

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java [hunk 2]
```diff
@@ -177,8 +177,7 @@ public String getFile(String relpath) {
     public void deleteFile(String filePath) {
         try {
             // Delete file
-            String sqlsearch = "delete from assets where filePath= ?;";
-            database.rawQuery(sqlsearch, new String[] { filePath }).moveToFirst();;
+            database.delete(DATABASE_TABLE_NAME, "filePath like ?", new String[] { filePath + '%' });
```
Comment (ronkorving) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: Why do we need LIKE here?

Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: This is to allow deleting directories.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java (hunk 1 above)
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: Sounds small yeah, but It does the job... and we use a buffer of the same size for downloading:
https://github.com/Wizcorp/phonegap-plugin-wizAssets/pull/10/files#diff-6d5265fda5622042ce0f322512001a31L243
Doing benchmarking would be great here. I can keep it as a TODO for a later PR.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java [hunk 3]
```diff
@@ -39,58 +47,82 @@
...
-                    for (int i=0; i<splitURL.length-1; i++) {
-                        dirName = dirName+splitURL[i]+"/";
+                    for (int i=0; i < splitURL.length-1; i++) {
+                        dirName = dirName + splitURL[i] + "/";
+                    }
+                    String uri = dirName + fileName;
+                    String filePath = buildAssetFilePathFromUri(uri);
+                    String asset = getAssetFilePathFromUri(uri);
+                    File file = new File(filePath);
+                    if (asset != null && file.exists()) {
```
Comment (ronkorving) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: It's faster and less likely to suffer from race conditions if you simply setLastModified and don't do an exists check. If setting last modified fails because the file does not exist, you can move to your else block.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java [hunk 4]
```diff
@@ -39,58 +47,82 @@
...
-                    String fileName = splitURL[splitURL.length-1];
-
-                    // Build directory
                     String dirName = "";
-                    for (int i=0; i<splitURL.length-1; i++) {
-                        dirName = dirName+splitURL[i]+"/";
+                    for (int i=0; i < splitURL.length-1; i++) {
+                        dirName = dirName + splitURL[i] + "/";
+                    }
+                    String uri = dirName + fileName;
```
Comment (ronkorving) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: So `uri` is the same as `fileURL`? I don't see the point of all this string jumbling...

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java [hunk 5]
```diff
@@ -98,114 +130,225 @@ public void run() {
...
+            deleteFile(file);
+            // Delete from database
+            wizAssetManager.deleteFile(filePath);
+        }
+    }
+
+    private void deleteFile(File file) throws IOException {
+        if(file.isDirectory()) {
+            //directory is empty, then delete it
+            if(file.list().length == 0){
```
Comment (ronkorving) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: `file.list()` seems like an expensive directory traversal operation. You're doing it here, and on line 207. In the case of a non-empty folder, that means twice the cost. Actually three times, because you then repeat it on line 216. It seems to me that this function could be half the length and twice the speed.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java (hunk 2 above)
Comment (ronkorving) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: Ah, so you can delete all files in a particular folder in one go? :)
Why not create a `deleteFolder` for that instead? That would keep single file deletions faster, while at the same time you can ensure that your filePath ends with a slash.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java (hunk 1 above)
Comment (ronkorving) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: I can almost guarantee it's a very unfortunate size. Rather than benchmarking, I would suggest researching best practices. They must exist.

Comment (ronkorving) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: First Google hit shows where this original code came from: http://stackoverflow.com/questions/106770/standard-concise-way-to-copy-a-file-in-java
It also shows better alternatives.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java (hunk 3 above)
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: Right. But I forgot to remove the ```setLastModified``` since it should not be a part of this PR. I will remove it.
I will still need the exists check after that.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java (hunk 5 above)
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java]: Good catch, gonna modify this.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java (hunk 1 above)
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: Should it be part of this PR?

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java (hunk 2 above)
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: That seems do-able, I'll have a look.

File: platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java (hunk 1 above)
Comment (jrouault) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: I willing to modify this but I would prefer to do this in another PR.
I didn't take a deep look at those better alternatives but we may face some API limitations due to Android (upgrading ADT might solve some too).

Comment (ronkorving) [inline platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetManager.java]: Not in this PR, sorry. Do it whenever works for you, I just wanted to point out a flaw I noticed while reviewing.

**Comment by jrouault** [comment]
Updated:
- removed jumbling strings
- removed multiple calls to File.list
- added a specific SQL request to delete folder URI

File: platforms/ios/HelloCordova/Plugins/WizAssetsPlugin/WizAssetsPlugin.m [hunk 6]
```diff
@@ -37,18 +51,21 @@ - (void)backgroundDownload:(CDVInvokedUrlCommand *)command {
...
-            if ([filemgr createDirectoryAtPath:fullDir withIntermediateDirectories:YES attributes:nil error: NULL] == YES) {
-
+        NSError *error = nil;
+        NSData *urlData = [NSData dataWithContentsOfURL:url options:NSDataReadingUncached error:&error];
+
+        if (error) {
+            returnString = [NSString stringWithFormat:@"error - %@", error];
+        } else if (urlData) {
+            // Check if we didn't received a 401
+            // TODO: We might want to find another solution to check for this kind of error, and check other possible errors
```
Comment (ronkorving) [inline platforms/ios/HelloCordova/Plugins/WizAssetsPlugin/WizAssetsPlugin.m]: I think you really do.

Comment (jrouault) [inline platforms/ios/HelloCordova/Plugins/WizAssetsPlugin/WizAssetsPlugin.m]: We sure do. :)
```NSData``` is not the right way to download files in our case so it will be changed to use ```NSURLSession``` in a next PR.

Comment (ronkorving) [inline platforms/ios/HelloCordova/Plugins/WizAssetsPlugin/WizAssetsPlugin.m]: :+1:

**Comment by jrouault** [comment]
Updated:
- ensure path is directory when deleting folder in database
- minimize IO at folder creation

**Comment by jrouault** [comment]
Updated

**Comment by ronkorving** [comment]
LGTM
